File: common/src/main/java/cloudcmd/common/IndexStorage.java
package cloudcmd.common;

import org.json.JSONArray;
import org.json.JSONObject;

public interface IndexStorage
{
  // query
  //  find by tag
  //  find by other properties
  // init table
  // purge table
  // insert
  // flush
  // shutdown
  // add tag and batch
  // rm tag and batch

  void init();
  void purge();
  void flush();
  void shutdown();

  void add(JSONObject meta);

  JSONArray find(JSONObject filter);

  void addTag(JSONArray array, String[] tags);
  void removeTag(JSONArray array, String[] tags);
}

File: common/src/main/scala/cloudcmd/common/engine/IndexStorage.scala
package cloudcmd.common.engine

import cloudcmd.common.FileMetaData
import org.json.JSONArray
import org.json.JSONObject

trait IndexStorage extends EventSource  {
  def init(configRoot: String)

  def purge

  def flush

  def shutdown

  def reindex()

  def find(filter: JSONObject): Seq[FileMetaData]

  def add(fmd: FileMetaData)

  def addAll(fmds: Seq[FileMetaData])

  def remove(fmd: FileMetaData)

  def pruneHistory(fmds: Seq[FileMetaData])

  def get(fmds: Seq[FileMetaData])

  def ensure(fmds: Seq[FileMetaData], blockLevelCheck: Boolean)

  def remove(fmds: Seq[FileMetaData])

  def addTags(fmds: Seq[FileMetaData], tags: Set[String]): Seq[FileMetaData]
}